### PR TITLE
campaigns: Cache external changeset diffstats

### LINKED ISSUE #11075: Cache diffstat for external changesets
Previously, we introduced caching on the diffstats of patches. That was relatively simple, as we know when the diff changes.

For ExternalChangesets however, we need to have a more complex logic, hence it hasn't been implemented yet.
@mrnugget and I just brainstormed in the watercooler and came up with the following ideas:

- We store the diffstat on the changeset, like we do with patch
- We need to recalculate it when either baseOID (base branch change) or headOID (someone pushed to the branch) change
	- We (soon #11000) rely on `base` and `head` in the `Diff` resolver on ExternalChangeset
	- For GitHub, the head only get's updated once we sync events, so we need to check whether the headOID changed when syncing and then recalculate
	- For Bitbucket, we currently cannot detect the headOID changed from events we receive. Hence, we should store the OID of the commit we generate with gitserver when running the changesetjob, and on every received event use `git resolve-rev` to find out if the head refs `HEAD` changed. Recalculate if so.

### PR DESCRIPTION
**Pull request by LawnGnome**
Fixes #11075.

The general approach here is to cache the changeset diffstat if the base or head OID has changed when we sync the changeset, then use that cached changeset when a GraphQL query is made that requires a campaign or changeset diffstat.

There is one observable behaviour change: closed changesets previously returned zero diffstats, but now return their actual diffstat. I think this is OK, and it simplifies the implementation, but we could force the syncer to cache zeroes for closed changesets to return to the old behaviour. I've kept the relevant test change in a separate commit for now to make it easier to change this if necessary.

Since campaign diffstats require knowledge of which changesets are accessible to the user, I've elected to use the simple approach for now of continuing to use the GraphQL connection. (See the discussion below between @mrnugget and I for more detail.) This is technically suboptimal, as we're pulling in sync data that we don't need to, but I think that's likely OK pending further work on being able to resolve permissions at the store level. I'm also happy to attempt @mrnugget's suggested optimisation in a future PR.

### CODE REVIEW

File: enterprise/internal/campaigns/resolvers/campaigns.go [hunk 1]
```diff
@@ -290,21 +290,24 @@ func (r *campaignResolver) DiffStat(ctx context.Context) (*graphqlbackend.DiffSt
...
 	}
 
 	totalStat := &graphqlbackend.DiffStat{}
-
-	for _, repoComp := range repoComparisons {
-		fileDiffs := repoComp.FileDiffs(&graphqlbackend.FileDiffsConnectionArgs{})
-		s, err := fileDiffs.DiffStat(ctx)
-		if err != nil {
-			return nil, err
+	for _, cs := range changesets {
```
Comment (mrnugget) [inline enterprise/internal/campaigns/resolvers/campaigns.go:299]: You can skip everything in this method up until this point, by simply adding a "GetChangesetsDiffStat" method to the `Store`. The only reason why I constructed the `changesetsConnection` and iterated over the `Nodes` was because we needed the `repoComparison`. Since we don't need that anymore and have the fields cached in the database, we can simply use that.

If we write a migration like this:

https://github.com/sourcegraph/sourcegraph/blob/9d1d9dec2895a6040bf671fcdee1570df6c7f273/enterprise/internal/campaigns/migrations.go#L9-L37

Then it's totally fine.

And I'm not even sure whether we need a migration, because we _could_ change our syncer to give higher priority to changesets without a diffstat. Then they'd be synced on startup (if I'm not mistaken)

File: enterprise/internal/campaigns/resolvers/changesets.go [hunk 2]
```diff
@@ -411,6 +412,17 @@ func (r *changesetResolver) Diff(ctx context.Context) (*graphqlbackend.Repositor
 	})
 }
 
+func (r *changesetResolver) DiffStat(ctx context.Context) (*graphqlbackend.DiffStat, error) {
+	if r.DiffStatAdded == nil || r.DiffStatChanged == nil || r.DiffStatDeleted == nil {
```
Comment (mrnugget) [inline enterprise/internal/campaigns/resolvers/changesets.go:416]: See previous comment regarding `changeset.DiffStat()` method :)

File: enterprise/internal/campaigns/state.go [hunk 3]
```diff
@@ -388,6 +400,91 @@ func computeReviewState(statesByAuthor map[string]campaigns.ChangesetReviewState
 	return selectReviewState(states)
 }
 
+func computeDiffStat(ctx context.Context, c *cmpgn.Changeset) (*diff.Stat, error) {
+	repo, err := db.Repos.Get(ctx, c.RepoID)
```
Comment (mrnugget) [inline enterprise/internal/campaigns/state.go:404]: Is it possible to pass the repository in here? I believe that in a lot of places where we call `SetDerivedState` we already have the repo at hand.

File: enterprise/internal/campaigns/state.go [hunk 4]
```diff
@@ -388,6 +400,91 @@ func computeReviewState(statesByAuthor map[string]campaigns.ChangesetReviewState
...
+		}
+
+		c.SetDiffStat(stat)
+		c.SyncState = *newState
+	}
+
+	return nil, nil
+}
+
+func computeSyncState(ctx context.Context, c *cmpgn.Changeset, repo gitserver.Repo) (*cmpgn.ChangesetSyncState, error) {
```
Comment (mrnugget) [inline enterprise/internal/campaigns/state.go:447]: Ah, I now get what you're doing with the new column. If we can do that in a central place (like here, in `SetDerivedState`) that would be great.

File: enterprise/internal/campaigns/resolvers/campaigns.go (hunk 1 above)
Comment (LawnGnome) [reply enterprise/internal/campaigns/resolvers/campaigns.go:299]: I think the priority approach is the better one here: unlike the patch migration, a changeset migration might block for a significant period of time talking to external providers, and we can just prefill the priority queue when starting the syncer.

Comment (LawnGnome) [reply enterprise/internal/campaigns/resolvers/campaigns.go:299]: > You can skip everything in this method up until this point, by simply adding a "GetChangesetsDiffStat" method to the `Store`. The only reason why I constructed the `changesetsConnection` and iterated over the `Nodes` was because we needed the `repoComparison`.

I'm not sure this works: we need to be able to check if the user has access to the changeset, and that seems to require setting up the connection. Is there another way to do that?

Comment (mrnugget) [reply enterprise/internal/campaigns/resolvers/campaigns.go:291]: Ah, good catch! Yes, we need to check for repository permissions. Since we're not able to JOIN against repo permissions yet (I've talked with @unknwon about this already), we'd need to do something like this here:

https://github.com/sourcegraph/sourcegraph/blob/f7f82762b677736d494ef4bc0e47c7e4e1322080/enterprise/internal/campaigns/resolvers/patch_sets.go#L80-L120

Meaning: fetch changesets, get repo IDs, filter out inaccessible repo IDs, then get diffstat only for accessible changesets. Again, since we can't do a JOIN we can't do that in a single SQL query. 

The only advantage this (fetch changesets, fetch repos, etc.) has over the "use the `changesetsConnection`" approach is that the `changesetsConnection` also loads the syncData, which we don't need. But it's also noisy.

If we put the logic into a separate function, we might be able to reuse it? `listAccessibleChangesets` or something.

In summary: I take back my original comment on "get rid of this" and leave the choice up to you.

**Comment by LawnGnome** [comment]
I ended up making a couple more cleanups along the way, but this is now ready for review.

File: enterprise/internal/campaigns/state.go [hunk 5]
```diff
@@ -388,6 +428,92 @@ func computeReviewState(statesByAuthor map[string]campaigns.ChangesetReviewState
...
+	}
+
+	rev, err := git.ResolveRevision(ctx, repo, nil, ref, nil)
+	return string(rev), err
+}
+
+// changesetGitserverRepo looks up a gitserver.Repo based on the RepoID within a
+// changeset.
+func changesetGitserverRepo(ctx context.Context, c *cmpgn.Changeset) (*gitserver.Repo, error) {
+	repo, err := db.Repos.Get(ctx, c.RepoID)
```
Comment (mrnugget) [inline enterprise/internal/campaigns/state.go:510]: Food for thought (and no need to change anything here): before we added support for repository permissions we solely used `repos.Store` to get to repositories in our `campaigns` packages. `db.Repos`, though, supports repository permissions and `repos.Store` doesn't. Now we have two ways of querying repositories in our packages.  I'm wondering whether we should switch to using only `db.Repos` (which relies on the global `dbconn.Global`) or if there's another way to get unify this abstraction (at least from the point of view of `campaigns`).

File: enterprise/internal/campaigns/resolvers/changesets.go [hunk 6]
```diff
@@ -406,6 +406,13 @@ func (r *changesetResolver) Diff(ctx context.Context) (*graphqlbackend.Repositor
 	})
 }
 
+func (r *changesetResolver) DiffStat(ctx context.Context) (*graphqlbackend.DiffStat, error) {
+	if stat := r.Changeset.DiffStat(); stat != nil {
+		return graphqlbackend.NewDiffStat(*stat), nil
+	}
```
Comment (mrnugget) [inline enterprise/internal/campaigns/resolvers/changesets.go:412]: The docstring in `schema.graphql` mentions that this is only returned when the changeset is open.

If we can reliably get the diffstat even when the changeset is merged/closed, we should do that and return it here. But then the docstring needs to be updated.

File: enterprise/internal/campaigns/syncer.go [hunk 7]
```diff
@@ -308,6 +308,11 @@ func (s *ChangesetSyncer) Run(ctx context.Context) {
 		s.queue.Upsert(sched...)
 	}
 
+	// Prioritize changesets without diffstats on startup.
```
Comment (mrnugget) [inline enterprise/internal/campaigns/syncer.go:311]: Should we add a note that this can be removed in the future, once all changesets have been migrated?

File: enterprise/internal/campaigns/syncer_test.go [hunk 8]
```diff
@@ -174,11 +176,84 @@ func TestChangesetPriorityQueue(t *testing.T) {
...
+			if tc.wantError && err == nil {
+				t.Error("expected an error; got nil")
+			} else if !tc.wantError && err != nil {
+				t.Errorf("got unexpected error: %+v", err)
+			}
+
+			// We have to wait for the goroutine to exit lest we fail after this
+			// function is complete.
+			close(syncer.priorityNotify)
+			wg.Wait()
```
Comment (mrnugget) [inline enterprise/internal/campaigns/syncer_test.go:244]: Just FYI and feel-free-to-ignore: in other tests we've used a `done` channel that's closed inside the goroutine instead of a waitgroup, because we can then do a `select { }` over the `done` channel and a timeout channel, ensuring that the tests don't run for a long time if something goes wrong:

https://sourcegraph.com/github.com/sourcegraph/sourcegraph@716ceb75498932353008cdbb22e500d4221d15cb/-/blob/cmd/gitserver/server/serverutil_test.go#L319-323

File: enterprise/internal/campaigns/testing/mock_github.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/repoupdater"
+	"github.com/sourcegraph/sourcegraph/internal/repoupdater/protocol"
+	"github.com/sourcegraph/sourcegraph/internal/vcs/git"
+)
+
+type MockedGitHubChangesetSyncState struct {
```
Comment (mrnugget) [inline enterprise/internal/campaigns/testing/mock_github.go:15]: Three things:

1. Introducing this `testing` package is a great idea!
2. Introducing mocks that wrap mocks that go together a lot is an even better idea!
3. What about this is `GitHub` specific? It looks to me like a left-over and that it can be removed now?

File: enterprise/internal/campaigns/state.go [hunk 10]
```diff
@@ -388,6 +428,92 @@ func computeReviewState(statesByAuthor map[string]campaigns.ChangesetReviewState
...
 }
 
+// computeDiffStat computes the up to date diffstat for the changeset, based on
+// the values in c.SyncState.
+func computeDiffStat(ctx context.Context, c *cmpgn.Changeset, repo gitserver.Repo) (*diff.Stat, error) {
+	iter, err := git.Diff(ctx, git.DiffOptions{
+		Repo: repo,
+		Base: c.SyncState.BaseRefOid,
+		Head: c.SyncState.HeadRefOid,
+	})
```
Comment (eseliger) [inline enterprise/internal/campaigns/state.go:438]: As @mrnugget mentioned in his comment [here](https://github.com/sourcegraph/sourcegraph/pull/11279/files?file-filters%5B%5D=.go&file-filters%5B%5D=.md&file-filters%5B%5D=.snap&file-filters%5B%5D=.sql&file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx#r443420561), the dependency on the merge state doesn't seem to be enforced, other than mentioned in the schema. This function is always called - right? While I think this is how it should be (I want to know how much impact my campaign is going to have in total, not how much it has left), I don't think this works for merged/non-open PRs. I tried the same for showing the diff for those and I was not able to figure out a way to do this: https://github.com/sourcegraph/sourcegraph/pull/11000#issuecomment-636247881. Basically, The problem is that using only OIDs and possibly deleted branches, we cannot reliably determine the diffstat for the PRs at the point of closing/merging (like our supported codehosts display it).
At least from what I can tell, it's not practicable to get the diff from git for a state that doesn't exist anymore. 
Maybe @mrnugget has some more ideas how this could work? 

The approach I am targetting for implementing the above linked pr is getting the diff from the codehost, not from git. That would also probably be faster as we can cache the diff in the db, rather than reach out to gitserver when rendering the diff in the client UI. Basic flow:
- Fetch patch from codehost: https://patch-diff.githubusercontent.com/raw/sourcegraph/sourcegraph/pull/11279.patch
- Strip patch headers (the `.diff` is not sufficient, as the diff is not meant to be applied and doesn't have all the required diff flags set)
- Feed the diff to `go-diff`
- Optional: Cache/store until oids changed or, when once merged, forever.
- Continue with existing logic

Not sure how else we could fix this.

Comment (mrnugget) [reply enterprise/internal/campaigns/state.go:438]: I think before we go down that rabbit hole again, we should simply exclude merged/closed changesets from the combined diffstats and not return their diffstat.

Comment (eseliger) [reply enterprise/internal/campaigns/state.go:438]: Yeah I didn't mean we definitely need to do it now, but need to ensure we don't calculate that here in case the PR is not open, because it will very likely fail

File: enterprise/internal/campaigns/syncer_test.go (hunk 8 above)
Comment (LawnGnome) [reply enterprise/internal/campaigns/syncer_test.go:244]: I like that better, particularly the timeout point. Done.

File: enterprise/internal/campaigns/testing/mock_github.go (hunk 9 above)
Comment (LawnGnome) [reply enterprise/internal/campaigns/testing/mock_github.go:15]: You are correct that it's a left over. Good catch!

File: enterprise/internal/campaigns/syncer.go (hunk 7 above)
Comment (LawnGnome) [reply enterprise/internal/campaigns/syncer.go:311]: Yep, done.

File: enterprise/internal/campaigns/state.go (hunk 10 above)
Comment (LawnGnome) [reply enterprise/internal/campaigns/state.go:438]: Or, alternatively, we could early return from `SetDerivedState` if the changeset is merged or closed and not update the diffstat or sync state.

I don't have particularly strong feelings on this, though. I'm happy to add the check to the GraphQL resolver if we think that's the right way to go.

File: enterprise/internal/campaigns/state.go (hunk 5 above)
Comment (LawnGnome) [reply enterprise/internal/campaigns/state.go:510]: My feeling is that we should try to unify by having only one repo implementation, but I agree that that's out of scope for this PR, and I think that's a bigger conversation (that I know has already been raised).

**Comment by LawnGnome** [comment]
> One "Hmm"-thought I had that I wanted to mention (and I'm curious to hear your thoughts): I'm not 100% sure about the name `ChangesetSyncState` and `sync_state`. I think I get what it's used for, something like an internal cache, but at the same time I'm wondering: "why isn't ExternalState and ExternalReviewState etc. in `sync_data`?" I'm curious to hear how you think about that column and what should go in and what not.

To me, this column is for state that needs to be persisted between syncs that is _internal_ to `enterprise/internal/campaigns`: fields that we're not surfacing via GraphQL, but rather fields that we need to compute the fields that are surfaced.

I'm open to better naming, of course.

Technically, we could move the external fields into this JSON blob, but I think having that separation between "here are things we know we need, and we're calling them out in the schema as explicit fields" and "here is a blob of state that is hidden behind the curtain" helps with understanding the data model.

File: enterprise/internal/campaigns/resolvers/changesets.go (hunk 6 above)
Comment (LawnGnome) [reply enterprise/internal/campaigns/resolvers/changesets.go:412]: (not ignoring this, just holding this open because it basically depends on what we decide in the next PR comment)

**Comment by mrnugget** [comment]
> To me, this column is for state that needs to be persisted between syncs that is _internal_ to `enterprise/internal/campaigns`:

That makes sense to me, thanks! (Would also be a great doc comment, in case that's not there yet)

File: enterprise/internal/campaigns/state.go (hunk 10 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/state.go:438]: > we could early return from `SetDerivedState` if the changeset is merged or closed and not update the diffstat or sync state.

The result would be that we keep the most recent diff stat, right? If that's possible, that sounds good to me.

Comment (eseliger) [reply enterprise/internal/campaigns/state.go:438]: There might be the small chance of keeping an outdated diffstat, but I guess it doesn't matter.

- Syncer syncs PR
- I update the PR
- I merge the PR
- Syncer syncs PR

(Less likely with webhooks in place)

Comment (LawnGnome) [reply enterprise/internal/campaigns/state.go:446]: I've implemented the early return in 3c38903: if the changeset is not open, _and_ was not open on the previous sync, then the syncer will early return. I figure we should try once to update the diffstat after the changeset is closed to address @eseliger's point: if it fails because gitserver no longer has the head available, then no harm no foul (and the previous diffstat will still be present), but it helps to address the problem Erik noted.

**Comment by LawnGnome** [comment]
> Would also be a great doc comment, in case that's not there yet

I added a more concise version of that explanation to the `SyncState` field.

File: enterprise/internal/campaigns/state.go (hunk 10 above)
Comment (mrnugget) [reply enterprise/internal/campaigns/state.go:446]: Sounds good!

**Comment by mrnugget** [comment]
Feel free to merge, btw.
